README.md
# seitz-symbol

[![Continuous Integration status][status-png]][status]
[![Hackage page (downloads and API reference)][hackage-png]][hackage]
[![Hackage-Deps][hackage-deps-png]][hackage-deps]

Read and Display Seitz Symbol

## Quickstart

Make new stack project and move to project directory.

```shell
% stack new symopRepl
% cd symopRepl
```

Edit your stack.yaml and set the following:

```
extra-deps:
- matrix-as-xyz-0.1.2.2@sha256:9ba89fbcc029fc769eac26d10460de093c8e6bbbcb60385a3990aadd3f2a27bb,2071
- symmetry-operations-symbols-0.0.2.1@sha256:4a443a1bd8f4ac155bb90906ee90f41f00ad7b50fd282a47498a00048124634c,2845
- location: https://github.com/narumij/seitz-symbol/archive/0.1.0.0.dev4.tar.gz
```

Edit your package.yaml and set the following:

```
dependencies:
- base >= 4.8 && < 5
- matrix-as-xyz
- seitz-symbol
- symmetry-operations-symbols
```

Then start repl.

```shell
% stack repl
```

Setup packages and load modules.

```haskell
repl> :m Data.Matrix.AsXYZ Data.Matrix.SeitzSymbol Data.Matrix.SymmetryOperationsSymbols
```

Use like below.

```haskell
repl> prettyXYZ <$> fromSeitzSymbolS "{ 2 010 | 1/2 1/2 1/2 }"
Right "-x+1/2,y+1/2,-z+1/2"

repl>  (liftError . fromSeitzSymbolS) "{ 2 010 | 1/2 1/2 1/2 }" >>= fromMatrix
Right " 2 (0,1/2,0) 1/4,y,1/4"
```

## References

A. Michael Glazer et al. Seitz symbols Acta Cryst. (2014). A70

ネスポロ マッシモ:日本結晶学会誌 59，210-222(2017).
https://www.jstage.jst.go.jp/article/jcrsj/59/5/59_210/_pdf


 [hackage]: http://hackage.haskell.org/package/seitz-symbol
 [hackage-png]: http://img.shields.io/hackage/v/seitz-symbol.svg
 [hackage-deps]: http://packdeps.haskellers.com/reverse/seitz-symbol
 [hackage-deps-png]: https://img.shields.io/hackage-deps/v/seitz-symbol.svg

 [status]: http://travis-ci.org/narumij/seitz-symbol?branch=master
 [status-png]: https://api.travis-ci.org/narumij/seitz-symbol.svg?branch=master


ChangeLog.md
# Changelog for seitz-symbol

## 0.1.0.0  -- 2020-08-??

* First version. Released on an unsuspecting world.

## Unreleased changes



test/doctests.hs
module Main (main) where

import Test.DocTest

main :: IO ()
main = doctest [
--  "src/Data/Matrix/SeitzSymbol.hs"
  ]


test/SetizSymbolSpec.hs
module SetizSymbolSpec where

import Test.Hspec

import Data.Matrix.SeitzSymbol
import Data.Matrix.AsXYZ (fromXYZ,prettyXYZ)
import Data.Matrix.SymmetryOperationsSymbols

spec :: Spec
spec = do

   describe "Data.Matrix.SeitzSymbol" $ do
      it "fromSeitzSymbol" $ do
        prettyXYZ <$> fromSeitzSymbolS "{ 1 | 0 0 0 }"
          `shouldBe` Right "x,y,z"
        prettyXYZ <$> fromSeitzSymbolS "{ 2 010 | 1/2 1/2 1/2 }"
          `shouldBe` Right "-x+1/2,y+1/2,-z+1/2"
        prettyXYZ <$> fromSeitzSymbolS "{ 3+ 111 | 1/2 1/2 1/2 }"
          `shouldBe` Right "z+1/2,x+1/2,y+1/2"
        prettyXYZ <$> fromSeitzSymbolS "{ -3+ 111 | 1/2 1/2 1/2 }"
          `shouldBe` Right "-z+1/2,-x+1/2,-y+1/2"
        prettyXYZ <$> fromSeitzSymbolS "{ m 100 | 0 0 0 }"
          `shouldBe`  Right "-x,y,z"
        ((liftError . fromSeitzSymbolS) "{ 2 010 | 1/2 1/2 1/2 }" >>= fromMatrix)
          `shouldBe` Right " 2 (0,1/2,0) 1/4,y,1/4"
        ((liftError . fromSeitzSymbolS) "{ 3+ 111 | 1/2 1/2 1/2 }" >>= fromMatrix)
          `shouldBe` Right " 3+(1/2,1/2,1/2) x,x,x"
        ((liftError . fromSeitzSymbolS) "{ -3+ 111 | 1/2 1/2 1/2 }" >>= fromMatrix)
          `shouldBe` Right "-3+ x,x,x; 1/4,1/4,1/4"
        ((liftError . fromSeitzSymbolS) "{ m 100 | 0 0 0 }" >>= fromMatrix)
          `shouldBe` Right " m  0,y,z"

      it "fromSeitzSymbolH" $ do
        prettyXYZ <$> fromSeitzSymbolHexS "{ m 100 | 0 0 0 }"
          `shouldBe` Right "y-x,y,z"
        prettyXYZ <$> fromSeitzSymbolHexS "{ m 120 | 0 0 0 }"
          `shouldBe` Right "x-y,-y,z"
        prettyXYZ <$> fromSeitzSymbolHexS "{ 2 100 | 0 0 0 }"
          `shouldBe` Right "x-y,-y,-z"
        ((liftError . fromSeitzSymbolHexS) "{ m 100 | 0 0 0 }" >>= fromMatrix)
          `shouldBe` Right " m  x,2x,z"
        ((liftError . fromSeitzSymbolHexS) "{ m 120 | 0 0 0 }" >>= fromMatrix)
          `shouldBe` Right " m  x,0,z"
        ((liftError . fromSeitzSymbolHexS) "{ 2 100 | 0 0 0 }" >>= fromMatrix)
          `shouldBe` Right " 2  x,0,0"

      it "toSeitzSymbol" $ do
        (toSeitzSymbolS . fromXYZ $ "x,y,z")
          `shouldBe` Just "{ 1 | 0 0 0 }"
        (toSeitzSymbolS . fromXYZ $ "-x+1/2,y+1/2,-z+1/2")
          `shouldBe` Just "{ 2 010 | 1/2 1/2 1/2 }"



src/Data/Matrix/SeitzSymbol/Table.hs
module Data.Matrix.SeitzSymbol.Table where

data Symbol
   = Id  --  '1'
   | T   --   t
   | Inv -- '-1'
   | M   --  'm'
   | A   --   a
   | B   --   b
   | C   --   c
   | D   --   d
   | G   --   g
   | N   --   n
   | R2  --  '2'
   | R3  --  '3'
   | R4  --  '4'
   | R6  --  '6'
   | RI3 -- '-3'
   | RI4 -- '-4'
   | RI6 -- '-6'
   deriving (Show,Eq)

data Lattice = Hexagonal | AnythingElse deriving (Eq)
type SymbolLabel = String
type Sense = String
type SymmetryElement = String
type Orientation = [Integer] -- orientation or location
type TransformedCoordinate = String
type AxisOrNormal = [Integer]

type Tbl = (Lattice,Symbol,SymbolLabel,Sense,SymmetryElement,Orientation,TransformedCoordinate,AxisOrNormal)

tbl :: [Tbl]
tbl = [
-- Table 11.2.2.1. Matrices for point-group symmetry operations and orientation
-- of corresponding symmetry elements, referred to a cubic, tetragonal, orthorhombic,
-- monoclinic, triclinic or rhombohedral coordinate system
  ( AnythingElse,  Id,  "1",  "",        "",         [],     "x,y,z", []),
  ( AnythingElse,  R2,  "2",  "",   "0,0,z", [ 0, 0, 1],   "-x,-y,z", []),
  ( AnythingElse,  R2,  "2",  "",   "0,y,0", [ 0, 1, 0],   "-x,y,-z", []),
  ( AnythingElse,  R2,  "2",  "",   "x,0,0", [ 1, 0, 0],   "x,-y,-z", []),
  ( AnythingElse,  R3,  "3", "+",   "x,x,x", [ 1, 1, 1],     "z,x,y", []),
  ( AnythingElse,  R3,  "3", "+", "x,-x,-x", [ 1,-1,-1],   "-z,-x,y", []),
  ( AnythingElse,  R3,  "3", "+", "-x,x,-x", [-1, 1,-1],   "z,-x,-y", []),
  ( AnythingElse,  R3,  "3", "+", "-x,-x,x", [-1,-1, 1],   "-z,x,-y", []),
  ( AnythingElse,  R3,  "3", "-",   "x,x,x", [ 1, 1, 1],     "y,z,x", []),
  ( AnythingElse,  R3,  "3", "-", "x,-x,-x", [ 1,-1,-1],   "-y,z,-x", []),
  ( AnythingElse,  R3,  "3", "-", "-x,x,-x", [-1, 1,-1],   "-y,-z,x", []),
  ( AnythingElse,  R3,  "3", "-", "-x,-x,x", [-1,-1, 1],   "y,-z,-x", []),
  ( AnythingElse,  R2,  "2",  "",   "x,x,0", [ 1, 1, 0],    "y,x,-z", []),
  ( AnythingElse,  R2,  "2",  "",   "x,0,x", [ 1, 0, 1],    "z,-y,x", []),
  ( AnythingElse,  R2,  "2",  "",   "0,y,y", [ 0, 1, 1],    "-x,z,y", []),
  ( AnythingElse,  R2,  "2",  "",  "x,-x,0", [ 1,-1, 0],  "-y,-x,-z", []),
  ( AnythingElse,  R2,  "2",  "",  "-x,0,x", [-1, 0, 1],  "-z,-y,-x", []),
  ( AnythingElse,  R2,  "2",  "",  "0,y,-y", [ 0, 1,-1],  "-x,-z,-y", []),
  ( AnythingElse,  R4,  "4", "+",   "0,0,z", [ 0, 0, 1],    "-y,x,z", []),
  ( AnythingElse,  R4,  "4", "+",   "0,y,0", [ 0, 1, 0],    "z,y,-x", []),
  ( AnythingElse,  R4,  "4", "+",   "x,0,0", [ 1, 0, 0],    "x,-z,y", []),
  ( AnythingElse,  R4,  "4", "-",   "0,0,z", [ 0, 0, 1],    "y,-x,z", []),
  ( AnythingElse,  R4,  "4", "-",   "0,y,0", [ 0, 1, 0],    "-z,y,x", []),
  ( AnythingElse,  R4,  "4", "-",   "x,0,0", [ 1, 0, 0],    "x,z,-y", []),
----
  ( AnythingElse, Inv, "-1",  "",   "0,0,0",         [],  "-x,-y,-z", []),
  ( AnythingElse,   M,  "m",  "",   "x,y,0", [ 0, 0, 1],    "x,y,-z", []),
  ( AnythingElse,   M,  "m",  "",   "x,0,z", [ 0, 1, 0],    "x,-y,z", []),
  ( AnythingElse,   M,  "m",  "",   "0,y,z", [ 1, 0, 0],    "-x,y,z", []),
  ( AnythingElse, RI3, "-3", "+",   "x,x,x", [ 1, 1, 1],  "-z,-x,-y", []),
  ( AnythingElse, RI3, "-3", "+", "x,-x,-x", [ 1,-1,-1],    "z,x,-y", []),
  ( AnythingElse, RI3, "-3", "+", "-x,x,-x", [-1, 1,-1],    "-z,x,y", []),
  ( AnythingElse, RI3, "-3", "+", "-x,-x,x", [-1,-1, 1],    "z,-x,y", []),
  ( AnythingElse, RI3, "-3", "-",   "x,x,x", [ 1, 1, 1],  "-y,-z,-x", []),
  ( AnythingElse, RI3, "-3", "-", "x,-x,-x", [ 1,-1,-1],    "y,-z,x", []),
  ( AnythingElse, RI3, "-3", "-", "-x,x,-x", [-1, 1,-1],    "y,z,-x", []),
  ( AnythingElse, RI3, "-3", "-", "-x,-x,x", [-1,-1, 1],    "-y,z,x", []),
  ( AnythingElse,   M,  "m",  "",  "x,-x,z", [ 1, 1, 0],   "-y,-x,z", []),
  ( AnythingElse,   M,  "m",  "",  "-x,y,x", [ 1, 0, 1],   "-z,y,-x", []),
  ( AnythingElse,   M,  "m",  "",  "x,y,-y", [ 0, 1, 1],   "x,-z,-y", []),
  ( AnythingElse,   M,  "m",  "",   "x,x,z", [ 1,-1, 0],     "y,x,z", []),
  ( AnythingElse,   M,  "m",  "",   "x,y,x", [-1, 0, 1],     "z,y,x", []),
  ( AnythingElse,   M,  "m",  "",   "x,y,y", [ 0, 1,-1],     "x,z,y", []),
  ( AnythingElse, RI4, "-4", "+",   "0,0,z", [ 0, 0, 1],   "y,-x,-z", []),
  ( AnythingElse, RI4, "-4", "+",   "0,y,0", [ 0, 1, 0],   "-z,-y,x", []),
  ( AnythingElse, RI4, "-4", "+",   "x,0,0", [ 1, 0, 0],   "-x,z,-y", []),
  ( AnythingElse, RI4, "-4", "-",   "0,0,z", [ 0, 0, 1],   "-y,x,-z", []),
  ( AnythingElse, RI4, "-4", "-",   "0,y,0", [ 0, 1, 0],   "z,-y,-x", []),
  ( AnythingElse, RI4, "-4", "-",   "x,0,0", [ 1, 0, 0],   "-x,-z,y", []),
-- Table 11.2.2.2. Matrices for point-group symmetry operations and orientation
-- of corresponding symmetry elements, referred to a hexagonal coordinate system
  (    Hexagonal,  Id,  "1",  "",        "",         [],     "x,y,z", []),
  (    Hexagonal,  R3,  "3", "+",   "0,0,z", [ 0, 0, 1],  "-y,x-y,z", []),
  (    Hexagonal,  R3,  "3", "-",   "0,0,z", [ 0, 0, 1],  "y-x,-x,z", []),
  (    Hexagonal,  R2,  "2",  "",   "0,0,z", [ 0, 0, 1],   "-x,-y,z", []),
  (    Hexagonal,  R6,  "6", "+",   "0,0,z", [ 0, 0, 1],   "x-y,x,z", []),
  (    Hexagonal,  R6,  "6", "-",   "0,0,z", [ 0, 0, 1],   "y,y-x,z", []),
  (    Hexagonal,  R2,  "2",  "",   "x,x,0", [ 1, 1, 0],    "y,x,-z", []),
  (    Hexagonal,  R2,  "2",  "",   "x,0,0", [ 1, 0, 0], "x-y,-y,-z", []),
  (    Hexagonal,  R2,  "2",  "",   "0,y,0", [ 0, 1, 0], "-x,y-x,-z", []),
  (    Hexagonal,  R2,  "2",  "",  "x,-x,0", [ 1,-1, 0],  "-y,-x,-z", []),
--
  (    Hexagonal,  R2,  "2",  "",  "x,2x,0", [ 1, 2, 0],  "y-x,y,-z", []),
  (    Hexagonal,  R2,  "2",  "",  "2x,x,0", [ 2, 1, 0],  "x,x-y,-z", []),
-- 
  (    Hexagonal, Inv, "-1",  "",   "0,0,0",         [],  "-x,-y,-z", []),
  (    Hexagonal, RI3, "-3", "+",   "0,0,z", [ 0, 0, 1],  "y,y-x,-z", []),
  (    Hexagonal, RI3, "-3", "-",   "0,0,z", [ 0, 0, 1],  "x-y,x,-z", []),
  (    Hexagonal,   M,  "m",  "",   "x,y,0", [ 0, 0, 1],    "x,y,-z", []),
  (    Hexagonal, RI6, "-6", "+",   "0,0,z", [ 0, 0, 1], "y-x,-x,-z", []),
  (    Hexagonal, RI6, "-6", "-",   "0,0,z", [ 0, 0, 1], "-y,x-y,-z", []),
  (    Hexagonal,   M,  "m",  "",  "x,-x,z", [ 1, 1, 0],   "-y,-x,z", []),
-- 以下二つが、Orientationを利用して解の補正をすることができなかったため、代替値を用意している
-- 行列を解いた場合の解平面とorientationが一致していない可能性（2017の試行錯誤)
-- そもそも勘違いの可能性もまだあるので、のちのち再確認する。
-- hackageに登録するに至らない理由の一つ
-- 正規化された値として正しいが、正規化の結果、復元に必要な情報が欠落してしまった可能性(2020リファクタリング時の見解)
-- どうしてこうなっているのか、やっぱりわからない。
  (    Hexagonal,   M,  "m",  "",  "x,2x,z", [ 1, 0, 0],   "y-x,y,z", [ 2,-1, 0]),
  (    Hexagonal,   M,  "m",  "",  "2x,x,z", [ 0, 1, 0],   "x,x-y,z", [-1, 2, 0]),

  (    Hexagonal,   M,  "m",  "",   "x,x,z", [ 1,-1, 0],     "y,x,z", []),
  (    Hexagonal,   M,  "m",  "",   "x,0,z", [ 1, 2, 0],  "x-y,-y,z", []),
  (    Hexagonal,   M,  "m",  "",   "0,y,z", [ 2, 1, 0],  "-x,y-x,z", [])
-- Notice
-- Hexagonal用のテーブルが、HexagonalのITで出現する対称操作全てをカバーしているわけではないことに注意
-- hexagonalでのlookup時には、hexagonal部分が優先となるよう、順番をいれかえている
-- それ以外のケースでは除外している
  ]


src/Data/Matrix/SeitzSymbol/Parser.hs
{- |
Module      : Data.Matrix.SeitzSymbol.Parser
Copyright   : (c) Jun Narumi 2020
License     : MIT
Maintainer  : narumij@gmail.com
Stability   : experimental

Seitz Symbol parser and etc.

[References]

A. Michael Glazer et al. Seitz symbols Acta Cryst. (2014). A70

ネスポロ マッシモ:日本結晶学会誌 59，210-222(2017).
https://www.jstage.jst.go.jp/article/jcrsj/59/5/59_210/_pdf

-}
module Data.Matrix.SeitzSymbol.Parser (
  SeitzSymbol(..),
  seitzSymbol,
  toMatrix,
  toSeitzSymbol,
  toString,
  ) where

import Data.Ratio (Ratio(..),(%))
import Text.Parsec
import Text.Parsec.String (Parser)

import Data.Ratio.Slash (Slash(..))
import Data.Matrix (Matrix(..),fromLists,toList,submatrix)
import Data.Matrix.AsXYZ (fromXYZ)
import qualified Data.Matrix as M ((<->),(<|>))

import Data.Matrix.SymmetryOperationsSymbols.Common (properMatricesForPointGroup,MatrixForPointGroupCorrespondingSymmetryElement(..))

type SeitzSymbol a = (String,String,(a,a,a),(Ratio a,Ratio a,Ratio a))

optionSpaces :: Parser ()
optionSpaces = skipMany space

identity :: Parser String
identity = do
  char '1'
  return "1"

symbol :: Parser (String,String)
symbol = try irot <|> try rot <|> miller <|> two
  where
    sign = oneOf "-+"
    rot = do
      a <- oneOf "346"
      b <- sign
      return (a:[],b:[])
    two = do
      char '2'
      return ("2","")
    miller = do
      char 'm'
      return ("m","")
    irot = do
      char '-'
      a <- oneOf "346"
      b <- sign
      return ('-':a:[],b:[])

zero :: Num a => Parser a
zero = do
  char '0'
  return 0

one :: Num a => Parser a
one = do
  char '1'
  return 1

two :: Num a => Parser a
two = do
  char '2'
  return 2

minus :: Num a => Parser a
minus = do
  string "-1"
  return (-1)

d :: Num a => Parser a
d = do
  a <- (zero <|> one <|> two <|> minus)
  return a

orientation :: Num a => Parser (a,a,a)
orientation = try a <|> b
  where
    a = do
      char '['
      a <- d
      b <- d
      c <- d
      char ']'
      return (a,b,c)
    b = do
      a <- d
      b <- d
      c <- d
      return (a,b,c)


num :: (Num a, Read a) => Parser a
num = do
  x <- oneOf "123456789"
  xs <- many digit
  return $ read (x : xs)

int :: (Integral a, Read a) => Parser a
int = zero <|> num

fract :: (Integral a, Read a) => Parser (Ratio a)
fract = do
  n <- int
  char '/'
  d <- int
  return $ n % d

integer :: (Integral a, Read a) => Parser (Ratio a)
integer = do
  i <- int
  return (i%1)

number :: (Integral a, Read a) => Parser (Ratio a)
number = do
  try fract <|> integer

matrixPart :: Num a => Parser (String,String,(a,a,a))
matrixPart = try a <|> b
  where
    a = do
      (sy,si) <- symbol
      spaces
      o <- orientation
      return (sy,si,o)
    b = do
      s <- identity
      return (s,"",(0,0,0))

seitzSymbol :: (Integral a, Read a) => Parser (SeitzSymbol a)
seitzSymbol = do
  char '{'
  optionSpaces
  (sy,si,o) <- matrixPart
  optionSpaces
  char '|'
  optionSpaces
  p <- number
  spaces
  q <- number
  spaces
  r <- number
  optionSpaces
  char '}'
  return (sy,si,o,(p,q,r))

toMatrix :: (Integral a,Read a) =>
            [MatrixForPointGroupCorrespondingSymmetryElement a]
          -> SeitzSymbol a
          -> Maybe (Matrix (Ratio a))
toMatrix tbl (sy,si,(o1,o2,o3),(p,q,r)) = build p q r <$> result
  where
    transformCoordinate (_,_,symbolLabel,sense,_,orientation,transformedCoordinate,_)
      = ( (symbolLabel,sense,if null orientation then [0,0,0] else orientation), transformedCoordinate )
    result = lookup (sy,si,[o1,o2,o3]) $ map transformCoordinate tbl
    build p q r xyz = _W M.<|> _w M.<-> fromLists [[0,0,0,1]]
      where
        _W = submatrix 1 3 1 3 . fromXYZ $ xyz
        _w = fromLists [[p],[q],[r]]

toString :: (Integral a, Show a) => SeitzSymbol a -> String
toString ("1",si,(o1,o2,o3),(p,q,r))
  = "{ " ++ "1"
  ++ " | "
  ++ show (Slash p) ++ " " ++ show (Slash q) ++ " " ++ show (Slash r)
  ++ " }"
toString (sy,si,(o1,o2,o3),(p,q,r))
  = "{ " ++ sy ++ si ++ " "
  ++ show o1 ++ show o2 ++ show o3
  ++ " | "
  ++ show (Slash p) ++ " " ++ show (Slash q) ++ " " ++ show (Slash r)
  ++ " }"

toSeitzSymbol :: Integral a => Matrix (Ratio a) -> Maybe (SeitzSymbol a)
toSeitzSymbol m = lookup w $ map tt properMatricesForPointGroup
  where
    getW = submatrix 1 3 1 3
    getw = submatrix 1 3 4 4
    w = getW m
    p:q:r:[] = toList . getw $ m
    tt (_,_,symbolLabel,sense,_,(o1:o2:o3:_),transformedCoordinate,_)
      = (getW . fromXYZ $ transformedCoordinate, (symbolLabel,sense,(o1,o2,o3),(p,q,r)))
    tt (_,_,symbolLabel,sense,_,[],transformedCoordinate,_)
      = (getW . fromXYZ $ transformedCoordinate, (symbolLabel,sense,(0,0,0),(p,q,r)))



src/Data/Matrix/SeitzSymbol.hs
{- |
Module      : Data.Matrix.SeitzSymbol
Copyright   : (c) Jun Narumi 2020
License     : MIT
Maintainer  : narumij@gmail.com
Stability   : experimental

Read and Display Seitz Symbol

[References]

A. Michael Glazer et al. Seitz symbols Acta Cryst. (2014). A70

ネスポロ マッシモ:日本結晶学会誌 59，210-222(2017).
https://www.jstage.jst.go.jp/article/jcrsj/59/5/59_210/_pdf

-}
module Data.Matrix.SeitzSymbol (
  P.SeitzSymbol(..),
  fromSeitzSymbolS,
  fromSeitzSymbolHexS,
  toSeitzSymbolS,
  ) where

import Text.Parsec
import Text.Parsec.String (Parser)

import Data.Ratio (Ratio(..))
import Data.Matrix (Matrix(..))
import Data.Matrix.AsXYZ (fromXYZ,prettyXYZ)
import qualified Data.Matrix.SeitzSymbol.Parser
  as P (SeitzSymbol(..),seitzSymbol,toMatrix,toSeitzSymbol,toString)
import Data.Matrix.SymmetryOperationsSymbols
import Data.Matrix.SymmetryOperationsSymbols.Common (
  properMatricesForPointGroup,
  hexagonalMatricesForPointGroup,
  MatrixForPointGroupCorrespondingSymmetryElement(..)
  )

parser :: (Integral a, Read a) => 
          [MatrixForPointGroupCorrespondingSymmetryElement a]
        -> Parser (Matrix (Ratio a))
parser tbl = do
  s <- P.seitzSymbol
  case P.toMatrix tbl s of
    Just m -> return m
    Nothing -> parserFail "Matrix not found."

-- | for cubic, tetragonal, orthorhombic, monoclinic and triclinic crystal systems.
--
-- >>> prettyXYZ <$> fromSeitzSymbolS "{ 1 | 0 0 0 }"
-- Right "x,y,z"
-- >>> prettyXYZ <$> fromSeitzSymbolS "{ 2 010 | 1/2 1/2 1/2 }"
-- Right "-x+1/2,y+1/2,-z+1/2"
-- >>> prettyXYZ <$> fromSeitzSymbolS "{ 3+ 111 | 1/2 1/2 1/2 }"
-- Right "z+1/2,x+1/2,y+1/2"
-- >>> prettyXYZ <$> fromSeitzSymbolS "{ -3+ 111 | 1/2 1/2 1/2 }"
-- Right "-z+1/2,-x+1/2,-y+1/2"
-- >>> prettyXYZ <$> fromSeitzSymbolS "{ m 100 | 0 0 0 }"
-- Right "-x,y,z"
-- >>> (liftError . fromSeitzSymbolS) "{ 2 010 | 1/2 1/2 1/2 }" >>= fromMatrix
-- Right " 2 (0,1/2,0) 1/4,y,1/4"
-- >>> (liftError . fromSeitzSymbolS) "{ 3+ 111 | 1/2 1/2 1/2 }" >>= fromMatrix
-- Right " 3+(1/2,1/2,1/2) x,x,x"
-- >>> (liftError . fromSeitzSymbolS) "{ -3+ 111 | 1/2 1/2 1/2 }" >>= fromMatrix
-- Right "-3+ x,x,x; 1/4,1/4,1/4"
-- >>> (liftError . fromSeitzSymbolS) "{ m 100 | 0 0 0 }" >>= fromMatrix
-- Right " m  0,y,z"
--
fromSeitzSymbolS :: (Integral a, Read a) =>
                    SourceName
                 -> Either ParseError (Matrix (Ratio a))
fromSeitzSymbolS s = parse (parser properMatricesForPointGroup) s s


-- | for hexagonal and trigonal crystal systems.
--
-- >>> prettyXYZ <$> fromSeitzSymbolHexS "{ m 100 | 0 0 0 }"
-- Right "y-x,y,z"
-- >>> prettyXYZ <$> fromSeitzSymbolHexS "{ m 120 | 0 0 0 }"
-- Right "x-y,-y,z"
-- >>> prettyXYZ <$> fromSeitzSymbolHexS "{ 2 100 | 0 0 0 }"
-- Right "x-y,-y,-z"
-- >>> (liftError . fromSeitzSymbolHexS) "{ m 100 | 0 0 0 }" >>= fromMatrix
-- Right " m  x,2x,z"
-- >>> (liftError . fromSeitzSymbolHexS) "{ m 120 | 0 0 0 }" >>= fromMatrix
-- Right " m  x,0,z"
-- >>> (liftError . fromSeitzSymbolHexS) "{ 2 100 | 0 0 0 }" >>= fromMatrix
-- Right " 2  x,0,0"
--
fromSeitzSymbolHexS :: (Integral a, Read a) =>
                       String
                    -> Either ParseError (Matrix (Ratio a))
fromSeitzSymbolHexS s = parse (parser hexagonalMatricesForPointGroup) s s

-- |
--
-- >>> toSeitzSymbolS . fromXYZ $ "x,y,z"
-- Just "{ 1 | 0 0 0 }"
-- >>> toSeitzSymbolS . fromXYZ $ "-x+1/2,y+1/2,-z+1/2"
-- Just "{ 2 010 | 1/2 1/2 1/2 }"
--
toSeitzSymbolS :: (Integral a, Show a) => Matrix (Ratio a) -> Maybe String
toSeitzSymbolS m = P.toString <$> P.toSeitzSymbol m
